### PR TITLE
doc: add performance tuning doc

### PR DESCRIPTION
**Pull request by Qard**
@watson I'm not sure this is quite done yet, but I don't have any more time to work on this tonight. Feel free to make your own edits, otherwise I'll get back to it next Tuesday.

### CODE REVIEW

**Comment by watson** [comment]
I added a link from the index page to the performance doc and fixed the commit message linter bug. With the new link, the index page will look like this:

image

I'm not sure if that's the best location for it?

I also added line breaks after punctuation marks. I always feels that it's a little excessive when ever I break on all punctuation marks, but at least it saves me from considering each case individually 😅

I also went over the docs to see what other config options could be used to tune the performance. We should consider adding some of these:

- [`flushInterval`](https://www.elastic.co/guide/en/apm/agent/nodejs/current/agent-api.html#flush-interval)
- [`maxQueueSize`](https://www.elastic.co/guide/en/apm/agent/nodejs/current/agent-api.html#max-queue-size)
- [`serverTimeout`](https://www.elastic.co/guide/en/apm/agent/nodejs/current/agent-api.html#server-timeout) - Only really an issue if your APM Server or the connection to it is unstable
- [`transactionMaxSpans`](https://www.elastic.co/guide/en/apm/agent/nodejs/current/agent-api.html#transaction-max-spans)
- [`ignoreUrls`](https://www.elastic.co/guide/en/apm/agent/nodejs/current/agent-api.html#ignore-urls) - Not really that performance related
- [`ignoreUserAgents`](https://www.elastic.co/guide/en/apm/agent/nodejs/current/agent-api.html#ignore-user-agents) - Not really that performance related

**Comment by watson** [comment]
Oh just remembered that we should add a link to this page from the README ToC as well

File: docs/performance.asciidoc [hunk 1]
```diff
@@ -0,0 +1,153 @@
+[[performance-tuning]]
+
+ifdef::env-github[]
+NOTE: For the best reading experience,
+please view this documentation at https://www.elastic.co/guide/en/apm/agent/nodejs/current/performance.html[elastic.co]
```
Comment (watson) [inline docs/performance.asciidoc]: I messed up when I moved some things around and now the actual generated html file is called `performance-tuning.html` because of the new anchor tag 😅 

**Option 1:** Rename the link to point to `performance-tuning.html`
**Option 2:** Keep the link as it is and rename the generated file back to just `performance.html`

I vote for option 1. But in that case we should also consider renaming the asciidoc file to match the generated html file. Just to make it easier to know what comes from what. All other doc files have a one-to-one filename correlation (`setup.asciidoc` being the only exception, as it generates `advanced-setup.html` - but we might consider "fixing" that as well).

What do you think?

File: docs/performance.asciidoc [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+require('elastic-apm-node').start({
+  transactionSampleRate: 0.2
+})
+----
+
+[float]
+[[performance-transaction-queue]]
+=== Transaction Queue
+
+The agent uses a queue to send to the apm server in batches.
```
Comment (watson) [inline docs/performance.asciidoc]: I find this paragraph a little hard to parse for some reason. How about something along the lines of this instead:

```asciidoc
The agent buffers the collected data using an in-memory queue before sending it to the APM Server.
The queue is flushed either after a specific <<performance-flush-interval,amount of time>> or when it reaches <<performance-max-queue-size,a certain size>> -
whichever comes first.
Lowering these defaults can reduce memory usage,
but will increase the number of requests to the APM Server.
```

File: docs/performance.asciidoc [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+This queue uses both a time interval and a size limit to balance memory usage against socket descriptor saturation.
+
+[float]
+[[performance-max-queue-size]]
```
Comment (watson) [inline docs/performance.asciidoc]: I would move this section below the `performance-flush-interval` section, so that they are listed in the same order as described above

File: docs/performance.asciidoc [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+The <<max-queue-size,`maxQueueSize`>> controls the maximum number of request transactions that may remain in the queue before they must be sent.
+
+Lowering this will reduce memory consumption,
+however it will increase the number of requests made to the apm server.
+
+[float]
+[[performance-flush-interval]]
+==== Flush Interval
+
+To prevent items from staying in the queue for a long time during low activity,
```
Comment (watson) [inline docs/performance.asciidoc]: Missing link to the config option

File: docs/performance.asciidoc [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+=== Transaction Queue
+
+The agent uses a queue to send to the apm server in batches.
+This queue uses both a time interval and a size limit to balance memory usage against socket descriptor saturation.
+
+[float]
+[[performance-max-queue-size]]
+==== Max Queue Size
+
+The <<max-queue-size,`maxQueueSize`>> controls the maximum number of request transactions that may remain in the queue before they must be sent.
```
Comment (watson) [inline docs/performance.asciidoc]: s/request transactions/transactions/

(all transactions are added to the queue, not only request transactions)

File: docs/performance.asciidoc [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+Lowering this will reduce memory consumption,
+however it will increase the number of requests made to the apm server.
```
Comment (watson) [inline docs/performance.asciidoc]: s/apm server/APM Server/

Comment (watson) [reply docs/performance.asciidoc]: Same goes for any other location in this doc where it says `apm server`

File: docs/performance.asciidoc [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+a flush interval is used to ensure the queue empties if anything in it is too old.
+
+Lowering the flush interval will ensure transactions are sent to the apm server faster,
+but may also result in increased HTTP traffic and cpu usage.
+
+[float]
+[[performance-server-timeout]]
+=== Server Timeout
+
+In the event that the connection to the apm server is slow or unstable,
```
Comment (watson) [inline docs/performance.asciidoc]: Missing link to the config option

File: docs/performance.asciidoc [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+
+To prevent items from staying in the queue for a long time during low activity,
+a flush interval is used to ensure the queue empties if anything in it is too old.
+
+Lowering the flush interval will ensure transactions are sent to the apm server faster,
+but may also result in increased HTTP traffic and cpu usage.
+
+[float]
+[[performance-server-timeout]]
+=== Server Timeout
```
Comment (watson) [inline docs/performance.asciidoc]: Consider renaming this headline to "APM Server Timeout"

File: docs/performance.asciidoc [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+This will mainly impact memory usage,
+but may also have a noticeable impact on speed too.
+The cpu time required to capture and convert a stack frame to something JavaScript can understand is not insignificant.
+
+[float]
+[[performance-error-log-stack-traces]]
+==== Error Log Stack Traces
+
+Most stack traces recorded by the agent will point to where the error was instantiated,
```
Comment (watson) [inline docs/performance.asciidoc]: Missing link to the config option

File: docs/performance.asciidoc (hunk 1 above)
Comment (Qard) [reply docs/performance.asciidoc]: Option 1 sounds good to me.

File: docs/performance-tuning.asciidoc [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+
+The first knob to reach for when tuning the performance of the agent is <<transaction-sample-rate,`transactionSampleRate`>>.
+Adjusting the sampling rate controls what percent of requests are traced.
+By default,
+the sample rate is set at `1.0`,
+meaning _all_ requests are traced.
+
+The sample rate will impact all four performance categories,
+so simply turning down the sample rate is an easy way to improve performance.
+
```
[the remark this answers is not in the document]
Comment (Qard) [reply docs/performance-tuning.asciidoc]: fixed.
